### PR TITLE
Link mentioned user in markdown only if they are visible to viewer

### PR DESCRIPTION
**Pull request by yardenshoham**
We need to make sure a user can't confirm the existence of a user with private visibility

* Follow up on #21533 

### Before

<details>

#### User
image

#### Admin
image

</details>

#### User
image

#### Admin
image

### CODE REVIEW

File: services/markup/processorhelper.go [hunk 1]
```diff
@@ -8,22 +8,24 @@ import (
...
-			// Only link if the user actually exists
-			userExists, err := user.IsUserExist(ctx, 0, username)
+			mentionedUser, err := user.GetUserByName(ctx, username)
 			if err != nil {
-				log.Error("Failed to validate user in mention %q exists, assuming it does", username)
-				userExists = true
+				return false
+			}
+
+			moduleCtx, ok := ctx.(*module_context.Context)
```
Comment (wxiaoguang) [inline services/markup/processorhelper.go]: It will fail the unit test.

The unit test should construct a &Context{}.

Comment (yardenshoham) [reply services/markup/processorhelper.go]: Like this? https://github.com/go-gitea/gitea/pull/21554/commits/08bdc8b9d55839b2b715db0ad5467c98d390a897

Comment (wxiaoguang) [reply services/markup/processorhelper.go]: Maybe you need to choose a public user and a private user, and fill the user to `Doer` to let the test check whether `IsUsernameMentionable -> IsUserVisibleToViewer` works as expected.

Comment (yardenshoham) [reply services/markup/processorhelper.go]: Something's wrong with DB access
```
code.gitea.io/gitea/modules/context.(*Context).Value(0xc001837ed0?, {0x17013e0?, 0x31975b0?})
        /workspace/gitea/modules/context/context.go:574 +0x5e
code.gitea.io/gitea/models/db.GetEngine({0x216d368, 0xc001a8aea0})
        /workspace/gitea/models/db/context.go:78 +0x57
```

Comment (wxiaoguang) [reply services/markup/processorhelper.go]: You need a mocked context, I did some fixes for it

**Review by wxiaoguang — APPROVED** [review]

**Review by lunny — APPROVED** [review]

File: services/markup/processorhelper_test.go [hunk 2]
```diff
@@ -6,15 +6,48 @@ package markup
...
+	unittest.AssertCount(t, &user.User{Name: userLimited}, 1)
+	unittest.AssertCount(t, &user.User{Name: userNoSuch}, 0)
+
+	// when using general context, use user's visibility to check
+	assert.True(t, ProcessorHelper().IsUsernameMentionable(context.Background(), userPublic))
+	assert.False(t, ProcessorHelper().IsUsernameMentionable(context.Background(), userLimited))
+	assert.False(t, ProcessorHelper().IsUsernameMentionable(context.Background(), userPrivate))
+	assert.False(t, ProcessorHelper().IsUsernameMentionable(context.Background(), userNoSuch))
+
+	// when using web context, use user's visibility to check
```
Comment (wxiaoguang) [inline services/markup/processorhelper_test.go]:
```suggestion
	// when using web context, use user.IsUserVisibleToViewer to check
```

File: services/markup/processorhelper.go [hunk 3]
```diff
@@ -8,22 +8,26 @@ import (
...
+
+			giteaCtx, ok := ctx.(*gitea_context.Context)
+			if !ok {
+				// when using general context, use user's visibility to check
+				return mentionedUser.Visibility.IsPublic()
 			}
-			return userExists
+
+			// when using gitea context (web context), use user's visibility and user's permission to check
+			return user.IsUserVisibleToViewer(giteaCtx, mentionedUser, giteaCtx.Doer)
```
Comment (lunny) [inline services/markup/processorhelper.go:30]: giteaCtx.Doer maybe nil which should be check out of `IsUserVisibleToViewer`

**Review by lunny — CHANGES_REQUESTED** [review]
https://github.com/go-gitea/gitea/pull/21554/files#r1002631586

**Comment by wxiaoguang** [comment]
> giteaCtx.Doer maybe nil which should be check out of IsUserVisibleToViewer
> https://github.com/go-gitea/gitea/pull/21554/files#r1002631586

No need. I have written the test to cover this case. `IsUserVisibleToViewer` can work with nil doer

File: services/markup/processorhelper.go (hunk 3 above)
Comment (wxiaoguang) [reply services/markup/processorhelper.go:30]: No need. See comment. Unless there are other problems.

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
make L-G-T-M work

File: services/markup/processorhelper.go [hunk 4]
```diff
@@ -8,22 +8,26 @@ import (
...
-			// Only link if the user actually exists
-			userExists, err := user.IsUserExist(ctx, 0, username)
+			mentionedUser, err := user.GetUserByName(ctx, username)
 			if err != nil {
-				log.Error("Failed to validate user in mention %q exists, assuming it does", username)
-				userExists = true
+				return false
+			}
+
+			giteaCtx, ok := ctx.(*gitea_context.Context)
```
Comment (wxiaoguang) [inline services/markup/processorhelper.go:23]: Need a fix:  Fix context usage #33554

**Pull request merged**